overlord/configstate/configmgr.go

```diff
@@ -22,23 +22,106 @@ package configstate
 import (
 	"regexp"
 
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/corecfg"
+	"github.com/snapcore/snapd/overlord/configstate/config"
 	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/snapstate"
 	"github.com/snapcore/snapd/overlord/state"
 )
 
 // ConfigManager is responsible for the maintenance of per-snap configuration in
 // the system state.
 type ConfigManager struct {
-	state *state.State
+	state  *state.State
+	runner *state.TaskRunner
 }
 
 // Manager returns a new ConfigManager.
-func Manager(s *state.State, hookManager *hookstate.HookManager) (*ConfigManager, error) {
+func Manager(st *state.State, hookManager *hookstate.HookManager) (*ConfigManager, error) {
+	// Most configuration is handled via the "configure" hook of the
+	// snaps. However some configuration is internally handled
+	hookManager.Register(regexp.MustCompile("^configure$"), newConfigureHandler)
+
+	// we handle core/snapd specific configuration internally because
+	// on classic systems we may need to configure things before any
+	// snap is installed.
+	runner := state.NewTaskRunner(st)
 	manager := &ConfigManager{
-		state: s,
+		state:  st,
+		runner: runner,
 	}
-
-	hookManager.Register(regexp.MustCompile("^configure$"), newConfigureHandler)
+	runner.AddHandler("run-core-configure", manager.doRunCoreConfigure, nil)
 
 	return manager, nil
 }
+
+// Ensure implements StateManager.Ensure.
+func (m *ConfigManager) Ensure() error {
+	m.runner.Ensure()
+	return nil
+}
+
+// Wait implements StateManager.Wait.
+func (m *ConfigManager) Wait() {
+	m.runner.Wait()
+}
+
+// Stop implements StateManager.Stop.
+func (m *ConfigManager) Stop() {
+	m.runner.Stop()
+}
+
+var corecfgRun = corecfg.Run
+
+func MockCorecfgRun(f func(tr corecfg.Conf) error) (restore func()) {
+	origCorecfgRun := corecfgRun
+	corecfgRun = f
+	return func() {
+		corecfgRun = origCorecfgRun
+	}
+}
+
+func (m *ConfigManager) doRunCoreConfigure(t *state.Task, tomb *tomb.Tomb) error {
+	var patch map[string]interface{}
+	var useDefaults bool
+
+	st := t.State()
+	snapName := "core"
+
+	// FIXME: lock taken for too long
+	st.Lock()
+	defer st.Unlock()
+
+	// FIXME: duplicated code from configureHandler.Before()
+	if err := t.Get("use-defaults", &useDefaults); err != nil && err != state.ErrNoState {
+		println(err.Error())
+		return err
+	}
+	if useDefaults {
+		var err error
+		patch, err = snapstate.ConfigDefaults(st, snapName)
+		if err != nil && err != state.ErrNoState {
+			return err
+		}
+	} else {
+		if err := t.Get("patch", &patch); err != nil && err != state.ErrNoState {
+			return err
+		}
+	}
+
+	tr := config.NewTransaction(st)
+	for key, value := range patch {
```

Comment:
transaction don't need the state lock until commit afaiu/r, so we could unlock from here until after runCoreCfg